src/components/Explore/actions/AddToFiltersAction.test.tsx

```diff
@@ -0,0 +1,117 @@
+import React from 'react';
+import { render, fireEvent } from '@testing-library/react'; 
+import { AddToFiltersAction, addToFilters } from './AddToFiltersAction';
+import { DataFrame } from '@grafana/data';
+import { AdHocFiltersVariable } from '@grafana/scenes';
+
+jest.mock('../../../utils/utils', () => ({
+  getFiltersVariable: jest.fn(),
+  getLabelValue: jest.fn(),
+}));
+
+const mockGetFiltersVariable = require('../../../utils/utils').getFiltersVariable;
+const mockGetLabelValue = require('../../../utils/utils').getLabelValue;
+
+describe('AddToFiltersAction', () => {
+  let variable: AdHocFiltersVariable;
```

Comment:
**[non-blocking]**: I personally find it more difficult to maintain tests in the long term when they use closures instead of factory functions.

For instance, something similar to:

```js
function buildAction(labels: Record<string, string> = { label1: 'value1', label2: 'value2' }) {
  const frame = {
    fields: [{ labels }],
  } as unknown as DataFrame;

  const setVariableState = jest.fn();
  const onClick = jest.fn();

  (getLabelValue as jest.Mock).mockReturnValue('value1');

  (getFiltersVariable as jest.Mock).mockReturnValue({
    state: { filters: [] },
    setState: setVariableState,
  } as unknown as AdHocFiltersVariable);

  return {
    action: new AddToFiltersAction({ frame, onClick }),
    onClick,
    setVariableState,
  };
}
```

would make the tests fairly easy to understand at first glance:

```js
  it('should add filter when no labelKey and exactly one label exists', () => {
    /// arrange
    const { action, onClick, setVariableState } = buildAction({ label1: 'value1' });

    // act
    action.onClick();

    // assert
    expect(setVariableState).toHaveBeenCalledWith({
      filters: [{ key: 'label1', operator: '=', value: 'value1' }],
    });
    expect(onClick).toHaveBeenCalledWith({ labelName: 'label1' });
  });
```

What do you think?

Reply:
Honestly I don't have a preference. I like the organisation of a factory function however I also like the clarity of the various bits laid out in the `beforeEach`.